packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,657 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
+        self.setDragPoint(self.points.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-drag-point') {
+        self.spliceDragPointIndex = null
+        self.setDragPoint((self.pathIsClosed) ? null : self.points.at(-1))
+        self.setClosePoint((self.pathIsClosed) ? null : self.points.at(0))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-close-point') {
+        self.splicePoints.length = 0
+        self.setClosePoint(null)
+        self.spliceReverse = false
+        self.spliceThroughBeginning = false
+        self.spliceActive = false
+        self.spliceClosePointIndex = null
+        self.drawActions.pop()
+      } else if (action.type == 'splice-append') {
+        let indexStart = self.splicePoints.length - action.pointsFromDrag.length
+        self.splicePoints.splice(indexStart, action.pointsFromDrag.length)
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.drawActions.pop()
+      } else if (action.type == 'splice-end') {
+        if (action.throughBeginning) {
+          let spliceIndex = Math.abs(action.pointIndexDrag - action.pointIndexClose)
+
+          for (let i = spliceIndex; i < self.points.length - 1; i++) {
+            self.splicePoints.push(self.points[i])
+          }
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, self.points.length - spliceIndex, ...action.pointsFromClose)
+          self.points.splice(0, 0, ...action.pointsFromDrag)
+        } else {
+          let spliceIndex = (action.isReversed)
+            ? action.pointIndexClose
+            : action.pointIndexDrag
+
+          for (let i = spliceIndex; i <= spliceIndex + action.pointsSpliceCount; i++) {
+            self.splicePoints.push(self.points[i])
+          }
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, action.pointsSpliceCount, ...action.pointsFromDrag)
+        }
+        self.spliceActive = true
+        self.spliceReverse = action.isReversed
+        self.spliceThroughBeginning = action.throughBeginning
+        self.spliceDragPointIndex = action.pointIndexDrag
+        self.spliceClosePointIndex = action.pointIndexClose
+
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.setClosePoint(self.points[self.spliceClosePointIndex])
+
+        self.drawActions.pop()
       }
+
+      self.redoActions.push(action)
     },
 
-    revertEdits() {
-      self.cancelClipPath()
-      self.dragPoint = null
-      self.targetPoint = null
+    redo() {
+      if (self.redoActions.length == 0) return
+
+      let action = toJS(self.redoActions.at(-1))
```

Comment:
```suggestion
      const action = self.redoActions.at(-1)
```

Same comment about `toJS` here, and you could define a view getter that returns `self.redoActions.at(-1)` too.